### PR TITLE
219 the fight function and the providers need a slight overhaul

### PR DESCRIPTION
**Pull request by lxgr-linux**
Just slightly extended the capabilities of the provider class.

### CODE REVIEW

File: pokete_classes/npcs.py [hunk 1]
```diff
@@ -203,7 +202,7 @@ class Trainer(NPC, Provider):
     def __init__(self, pokes, name, gender, texts, lose_texts,
                  win_texts):
         NPC.__init__(self, name, texts, side_trigger=False)
-        Provider.__init__(self, pokes)
+        Provider.__init__(self, pokes, False, 2)
```
Comment (mustafaquraish) [inline pokete_classes/npcs.py:205]: might be nice to use keyword args here so it's easy to see what `False` and `2` are referring to.

File: pokete_classes/npcs.py [hunk 2]
```diff
@@ -256,11 +255,49 @@ def action(self):
                 self.text({True: self.lose_texts,
                            False: self.win_texts + ["Here's $20!"]}
                           [is_winner])
-                heal(self) if is_winner else None
+                self.heal() if is_winner else None
```
Comment (mustafaquraish) [inline pokete_classes/npcs.py:258]: I think you should just do `if is_winner: self.heal()`, using the ternary if for non-expressions is not ideal

File: pokete_classes/providers.py [hunk 3]
```diff
@@ -27,12 +44,26 @@ def get_attack(self, fightmap, enem):
...
+        raise NotImplementedError
+
+
 class NatureProvider(Provider):
     """The Natures Provider
     ARGS:
         poke: One Pokete"""
     def __init__(self, poke):
-        super().__init__([poke])
+        super().__init__([poke], True, 1)
```
Comment (mustafaquraish) [inline pokete_classes/providers.py:66]: same point here with keyword args

File: pokete_classes/providers.py [hunk 4]
```diff
@@ -46,9 +77,50 @@ def get_attack(self, fightmap, enem):
...
+        ARGS:
+            fightmap: fightmap object"""
+        fightmap.outp.outp(f"A wild {self.curr.name} appeared!")
+
+    def handle_defeat(self, fightmap, winner):
+        """Function caleld when the providers current Pokete dies
+        ARGS:
+            fightmap: fightmap object
+            winner: the defeating provider"""
+        return False
```
Comment (mustafaquraish) [inline pokete_classes/providers.py:91]: What does this `False` represent? The docs don't mention the return value

Comment (lxgr-linux) [reply pokete_classes/providers.py:95]: It represents whether or not a Pokete was actually chosen, but I also updated the doc string.

File: pokete_classes/providers.py [hunk 5]
```diff
@@ -27,12 +44,28 @@ def get_attack(self, fightmap, enem):
             anem: The enemy Provider"""
         raise NotImplementedError
 
+    def greet(self, fightmap):
+        """Outputs a greeting text at the fights start:
+        ARGS:
+            fightmap: fightmap object"""
+        raise NotImplementedError
```
Comment (mustafaquraish) [inline pokete_classes/providers.py:51]: You're raising a `NotImplementedError` here, which makes me think this is an abstract method that needs to be overloaded by the child classes. This is a nice use case for the built-in `abc` module. Consider:

```py
from abc import ABC

class Provider(ABC):
    ...
    @abstractmethod
     def greet(self, fightmap):
         pass
```
If you define a child class that does *not* define the abstract method(s), then you get an error when trying to instantiate it, which is nice.

Comment (lxgr-linux) [reply pokete_classes/providers.py:51]: That's what I was trying to achieve.

**Review by mustafaquraish — APPROVED** [review]

**Pull request merged**